storage/src/vespa/storage/persistence/persistencethread.cpp

```diff
@@ -205,53 +203,66 @@ PersistenceThread::handleRemove(api::RemoveCommand& cmd, MessageTracker::UP trac
-                                   }
-                                   if (!response.wasFound()) {
-                                       metrics.notFound.inc();
-                                   }
-                                   tracker->sendReply();
-                               })));
+        auto task = makeResultTask([&metrics, &cmd, tracker = std::move(trackerUP)](spi::Result::UP responseUP) {
+            const spi::RemoveResult & response = dynamic_cast<const spi::RemoveResult &>(*responseUP);
+            if (tracker->checkForError(response)) {
+                tracker->setReply(std::make_shared<api::RemoveReply>(cmd, response.wasFound() ? cmd.getTimestamp() : 0));
+            }
+            if (!response.wasFound()) {
+                metrics.notFound.inc();
+            }
+            tracker->sendReply();
+        });
+        _spi.removeIfFoundAsync(bucket, spi::Timestamp(cmd.getTimestamp()), cmd.getDocumentId(), tracker.context(),
+                                std::make_unique<ResultTaskOperationDone>(*_sequencedExecutor, cmd.getBucketId(), std::move(task)));
     }
     return trackerUP;
 }
 
 MessageTracker::UP
-PersistenceThread::handleUpdate(api::UpdateCommand& cmd, MessageTracker::UP tracker)
+PersistenceThread::handleUpdate(api::UpdateCommand& cmd, MessageTracker::UP trackerUP)
 {
+    MessageTracker & tracker = *trackerUP;
     auto& metrics = _env._metrics.update[cmd.getLoadType()];
-    tracker->setMetric(metrics);
+    tracker.setMetric(metrics);
     metrics.request_size.addValue(cmd.getApproxByteSize());
 
-    if (tasConditionExists(cmd) && !tasConditionMatches(cmd, *tracker, tracker->context(), cmd.getUpdate()->getCreateIfNonExistent())) {
-        return tracker;
+    if (tasConditionExists(cmd) && !tasConditionMatches(cmd, tracker, tracker.context(), cmd.getUpdate()->getCreateIfNonExistent())) {
+        return trackerUP;
     }
-    
-    spi::UpdateResult response = _spi.update(getBucket(cmd.getUpdate()->getId(), cmd.getBucket()),
-                                             spi::Timestamp(cmd.getTimestamp()), cmd.getUpdate(), tracker->context());
-    if (tracker->checkForError(response)) {
-        auto reply = std::make_shared<api::UpdateReply>(cmd);
-        reply->setOldTimestamp(response.getExistingTimestamp());
-        tracker->setReply(std::move(reply));
+
+    spi::Bucket bucket = getBucket(cmd.getDocumentId(), cmd.getBucket());
+    if (_sequencedExecutor == nullptr) {
+        spi::UpdateResult response = _spi.update(bucket, spi::Timestamp(cmd.getTimestamp()), std::move(cmd.getUpdate()),tracker.context());
+        if (tracker.checkForError(response)) {
+            auto reply = std::make_shared<api::UpdateReply>(cmd);
+            reply->setOldTimestamp(response.getExistingTimestamp());
+            tracker.setReply(std::move(reply));
+        }
+    } else {
+        auto task = makeResultTask([&cmd, tracker = std::move(trackerUP)](spi::Result::UP responseUP) {
```

Comment:
Consider adding a comment stating why it's safe to refer to `cmd` as a reference capture to the argument given to the function itself, since the lifetime semantics aren't necessarily obvious.

Reply:
Fixed